rbd-mirror: Delete local image mirror when remote image mirroring is disabled

rjfd opened the pull request:
Ticket url: http://tracker.ceph.com/issues/14421

dillaman commented on src/librbd/internal.h:
```diff
@@ -86,9 +87,13 @@ namespace librbd {
   bool image_options_is_empty(rbd_image_options_t opts);
 
   int snap_set(ImageCtx *ictx, const char *snap_name);
+  int list_images_v2(librados::IoCtx& io_ctx,
```
Used?

rjfd replied:
Another leftover of the first implementation. I was using it in ImageDeleter.cc. Not anymore.

dillaman commented on src/include/Context.h:
```diff
@@ -463,6 +464,19 @@ class FunctionContext : public Context {
   boost::function<void(int)> m_callback;
 };
 
+class LambdaContext : public Context {
```
Minor: can we hold off on this for now and just use FunctionContext which also takes lambda?  I'd rather see a cleanup ticket to switch from FunctionContext / boost::function to the C++11 native.

rjfd replied:
Sure, no problem. 
Is the plan to substitute all occurrences of boost::function by C++11 native? I can do that afterwards :-)

dillaman replied:
There isn't a formulated plan for this in particular, but there has been talk about switching to C++11 primitives where available.  Such cleanup work should probably be left until the kraken release at this point.

dillaman commented on src/librbd/internal.cc:
```diff
@@ -1913,6 +1921,20 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force) {
 
+      if (force) {
+        C_SaferCond ctx;
+        ictx->exclusive_lock->shut_down(&ctx);
```
`exclusive_lock` might not be initialized yet -- perhaps move this down to be within the block near line 1939.  Probably should be holding a write lock against the `owner_lock` as well.

rjfd replied:
The reason I'm shutting down `exclusive_lock` here is because it needs to be done before the call to `ictx->operations->prepare_image_update()`

Why shouldn't the `exclusive_lock` be initialized at this point? Since ImageState::open() is synchronous, I though it would not be a problem to do this here.

dillaman replied:
If the exclusive lock feature is enabled, the `exclusive_lock` should be non-null -- but if an notify event comes in on the header while you have it open, there is a potential for a race if you aren't holding the lock.  If the exclusive lock feature isn't enabled (e.g. this wasn't a mirrored image, but just a regular ol' ```rbd rm foo```-call, it might be null)

rjfd replied:
> there is a potential for a race if you aren't holding the lock.
Which lock? the `owner_lock`?

dillaman replied:
Yeah -- ```owner_lock``` protects ```ImageCtx::exclusive_lock```.

dillaman commented on src/tools/rbd_mirror/ImageDeleter.cc:
```diff
@@ -0,0 +1,367 @@
+namespace mirror {
+
+ImageDeleter::ImageDeleter(string remote_cluster_name, RadosRef local_cluster)
+  : m_remote_name(remote_cluster_name),
```
Is there a reason to tie an image deleter to the remote cluster?  I was thinking that we would only need one for rbd-mirror.

rjfd replied:
You're right. In my first implementation I would only delete an non-primary image if the primary images was really deleted (and not only disabled) and to know that I would have to contact the remote cluster. That is why ImageDeleter was tied with a remote cluster.
Now we just need one ImageDeleter.

dillaman commented on src/tools/rbd_mirror/ImageDeleter.h:
```diff
@@ -0,0 +1,122 @@
+    FAILED_ON_REMOVE_IMAGE
+  };
+
+  class FailedDelete {
```
Might be nice to combine `FailedDelete` and `DeleteInfo` into a single structure.  That way you could avoid having both the Thread and SafeTimer both running / deleting at the same time -- and you could re-use the existing rbd-mirror SafeTimer to just move failed items from a  "retry" queue back into the main queue (to be processed when the main thread becomes available).

dillaman commented on src/tools/rbd_mirror/ImageDeleter.cc:
```diff
@@ -0,0 +1,367 @@
+
+  string header_oid = librbd::util::header_name(image_id);
+  ::SnapContext snapc;
+  int r = cls_client::get_snapcontext(ioctx, header_oid, &snapc);
```
Thinking about the case where the local has been offline and missed journal events for snap remove, perhaps the image deleter should effectively purge all existing snapshots on the image (assuming it has no active children first).

dillaman replied:
Never mind -- I see that you are doing that after invoking this helper function.

dillaman commented on src/tools/rbd_mirror/ImageDeleter.cc:
```diff
@@ -0,0 +1,367 @@
+    return;
+  }
+
+  switch (step) {
```
Why worry about tracking the state?  You should be able to just fall right through w/o worry (i.e. if mirroring has already been disabled, it's a no-op).

rjfd replied:
I was just trying to save some processing steps, but I agree that it may not worth the effort

dillaman commented on src/tools/rbd_mirror/ImageDeleter.cc:
```diff
@@ -0,0 +1,367 @@
+    *has_snapshots = true;
+    parent_spec pspec(ioctx->get_id(), image_id, snap_id);
+    set<pair<string, string> > names;
+    r = librbd::list_children(*ioctx, pspec, names);
```
The snap_unprotect operation will already check for children, so this might not be needed (and you might not need the new `list_children` method).

dillaman commented on src/tools/rbd_mirror/ImageDeleter.cc:
```diff
@@ -0,0 +1,367 @@
+  }
+
+failed_on_remove_image:
+  class SilentProgressContext : public ProgressContext {
```
`librbd::NoOpProgressContext` in `librbd/internal.h`

dillaman commented on src/tools/rbd_mirror/ImageDeleter.cc:
```diff
@@ -0,0 +1,367 @@
+    return ;
+  }
+
+  r = cls_client::mirror_image_remove(&ioctx, del_info->local_image_id);
```
Hmm -- we might have a race condition here.  If we crash after removing the local image from the local mirroring directory, we won't know that we need to delete the image when we recover.  The current `librbd::remove` method will disable mirroring -- but I think it needs to be broken up into two parts: (1) the current `mirror_image_disable_internal` call does everything except remove it from the directory, and (2) remove from the mirroring directory after removing from the image directory.

rjfd replied:
I think, I'll move the call of `mirror_image_remove` to after the call of  `librbd::remove`, and make necessary checks (check for -ENOENT) during all the execution of `process_image_delete`.

dillaman commented on src/tools/rbd_mirror/Replayer.cc:
```diff
@@ -220,6 +224,115 @@ int Replayer::init()
   return 0;
 }
 
+void Replayer::read_local_cluster_mirroring_info() {
```
Alternative: on start-up, you could populate a `map<pool_id, set<global image id> >` for all local images.  On the first call to `set_sources`, remove from the collection all global image ids from your local collection that exist in the remote.  Any images that remain in the collection are local images that no longer exist on the mirror.

dillaman commented on src/tools/rbd_mirror/Replayer.cc:
```diff
@@ -220,6 +224,115 @@ int Replayer::init()
+        if (r == -ENOENT) {
+          dout(20) << "remote image with global id " << it->second << "does "
+                   << "not exist. Scheduling delete of local mirror" << dendl;
+          m_image_deleter->schedule_image_delete(ioctx.get_id(), it->first,
```
Very important to not schedule any images for deletion if they are locally flagged as primary.  Need to guard here and also on shut-down (unless both deletion paths were combined in `set_sources` as noted above).

rjfd replied:
I'm going to check this condition inside `ImageDeleter::process_image_delete`

dillaman commented on src/tools/rbd_mirror/ImageDeleter.cc:
```diff
@@ -44,14 +46,85 @@ using namespace librbd;
+    std::string command;
+    int r;
+
+    command = "rbd image-deleter status "+cluster_name;
```
Minor: suggested alternative command naming: "rbd mirror deletion status"

dillaman commented on src/tools/rbd_mirror/Replayer.cc:
```diff
@@ -399,7 +403,17 @@ bool Replayer::stop_image_replayer(unique_ptr<ImageReplayer<> > &image_replayer)
+          }
+        }
+    );
+    image_replayer->stop(ctx);
```
We probably also want to guard to make sure we cannot create a new ImageReplayer if there is a pending delete scheduled for an image.  I am worried about the case where a user quickly deletes an image and recreates it.

rjfd replied:
I have a problem here. The information we keep for deleting an image is it's local_id and global_image_id.
If the user disables mirroring for the primary image and then re-enables it, the global_image_id will be different, so we will not be able check if there is a deletion scheduled. 
We could also keep the remote_id of the primary image, but it will also not work in the case where the user deletes and re-creates the primary image.

What will be the consequence of create a new ImageReplayer when there is a pending delete? Is because there still exists a local image with the same name? I that is the case, maybe what we should keep track in the DeleteInfo structure is the image name, which is common in the remote and local cluster.

dillaman replied:
Yeah, that was the case I was worried about -- I think we should match on remote image id or remote image name.

dillaman commented on src/tools/rbd_mirror/ImageDeleter.cc:
```diff
@@ -0,0 +1,367 @@
+      break;
+  }
+
+  mirror_image.global_image_id = del_info->global_image_id;
```
Probably good to ensure / guard that the local image is actually has the same global image id.  Getting a -EEXIST from `mirror_image_set` lets you know somehow another image id is now registered with that global id (bug?).  Getting a -EINVAL indicates the image now has a new global id (so somehow this image had mirroring disabled/re-enabled).

dillaman replied:
Yeah -- I think it can only be a bug for -EEXIST.  In such a case, log it and abort seems reasonable for now.  I think -EINVAL is possible if the user force-promotes the local image, disables mirroring, and re-enables it (e.g. crazy user).  Again, I think log and abort seems reasonable.

rjfd commented:
@dillaman finished the changes addressing your comments. I made big changes, so it will be as almost as the first time you are reviewing the code :stuck_out_tongue_closed_eyes:

rjfd commented:
Finished implementation and rebased on top of master

dillaman commented on src/test/rbd_mirror/test_ImageDeleter.cc:
```diff
@@ -0,0 +1,210 @@
+void register_test_rbd_mirror_image_deleter() {
+}
+
+class TestImageDeleter : public ::testing::Test {
```
If you inherit from the "TestFixture" class, it creates and connects to the two pools automatically.

rjfd replied:
I tried to use it, but it was easier to do the tests cleanup by deleting pools which goes against the logic of TestFixture.

dillaman replied:
Are you referring to the "FailedDeletePoolENOENT" test case where it deletes the local pool?  Could just create a pool specifically for that test case.

dillaman commented on src/test/rbd_mirror/test_ImageDeleter.cc:
```diff
@@ -0,0 +1,210 @@
+    m_local_image_id = replayer->get_local_image_id();
+    m_local_image_name = replayer->get_local_image_name();
+
+    if (seconds > 0) {
```
I'm probably missing something, but what's the purpose of the sleeps?  At a quick glance, it seems like it could be problematic/racy in the build jobs.

rjfd replied:
yeah, this was a nasty hack that I used to ensure (or not) that I wouldn't stop image replayer while it was still holding the local temporary snapshot. As a side effect, I found that if I stop during image replayer sync, the interruption handler does not remove the temporary snapshot, and the ImageDeleter will not be able to delete the local image because it fails while trying to delete the temporary snapshot with an -EROFS error. 

What do you suggest to ensure that the the sync is finished?

dillaman replied:
It would probably be better to test ImageDeleter in isolation from ImageReplayer -- and you could add a test case to ImageReplayer (or one of its subordinate classes) to verify it properly interacts with ImageDeleter.

rjfd replied:
That's great idea! Thanks

dillaman commented on src/tools/rbd_mirror/PoolWatcher.cc:
```diff
@@ -115,7 +116,13 @@ void PoolWatcher::refresh_images(bool reschedule)
       for (auto it = mirror_images.begin(); it != mirror_images.end(); ++it) {
-        image_ids.insert(ImageIds(it->first, it->second));
+        std::string image_name;
+        r = dir_get_name(&ioctx, RBD_DIRECTORY, it->first, &image_name);
```
I would recommend a single pass of retrieving all image names in the pool, followed by retrieving the mirrored images, and then cross-reference.  I'm not sure we can skip adding the mirrored image to the collection since that would be interpreted as a deleted image propagation.  Maybe a boost::optional<std::string> for the image name?

dillaman commented on src/tools/rbd_mirror/ImageDeleter.cc:
```diff
@@ -0,0 +1,491 @@
+  m_delete_queue_cond.Signal();
+}
+
+void ImageDeleter::is_image_scheduled_for_delete(const std::string& image_name,
```
Perhaps a name like ```wait_for_scheduled_deletion``` might be better?

rjfd replied:
agreed

dillaman commented on src/librbd/internal.cc:
```diff
@@ -1931,12 +1935,31 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force) {
-          return -EBUSY;
+        if (force) {
+          // releasing read lock to avoid a deadlock when upgrading to
+          // write lock in the stut_down process
```
Typo: "shut_down"

rjfd commented:
@dillaman Pushed the changes and rebased on top of master.

dillaman commented on src/librbd/internal.cc:
```diff
@@ -389,14 +390,16 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force) {
+    ldout(cct, 20) << "removed image state from rbd_mirroring object" << dendl;
+  }
 
   if (is_primary) {
```
Will need to delay this notification when ```!remove```.

rjfd replied:
Yes, I forgot about that. Thx!

dillaman commented:
@rjfd can you rebase and repush?

rjfd commented:
@dillaman done!

dillaman commented:
@rjfd Hit a race condition while manually testing.  I had an image replicated between two clusters and I disabled mirroring on the primary image.  The ```ImageDeleter``` properly picked up that it should delete the image, but when it tried to acquire the exclusive lock in order to delete it, it ended up blacklisting itself:

> 2016-05-03 09:13:25.883965 7f0bca7fc700 10 librbd::exclusive_lock::AcquireRequest: retrieved exclusive locker: client.4196@127.0.0.1:0/1749366697
> 2016-05-03 09:13:25.883969 7f0bca7fc700 10 librbd::exclusive_lock::AcquireRequest: send_get_watchers
> 2016-05-03 09:13:25.884277 7f0bca7fc700 10 librbd::exclusive_lock::AcquireRequest: handle_get_watchers: r=0
> 2016-05-03 09:13:25.884284 7f0bca7fc700 10 librbd::exclusive_lock::AcquireRequest: send_blacklist
> 2016-05-03 09:13:26.016784 7f0b6cff9700 10 librbd::exclusive_lock::AcquireRequest: handle_blacklist: r=0

The reason is because the ImageReplayer state machine was running on the same image -- and currently had the image open w/ the exclusive lock acquired.

dillaman commented on src/tools/rbd_mirror/ImageDeleter.cc:
```diff
@@ -0,0 +1,536 @@
+    return ;
+  }
+
+  dout(20) << "setted local image mirroring to disable" << dendl;
```
Minor: just saw this in the logs -- should be "set local image mirroring state to disabling"

dillaman commented:
@rjfd Correction: I am seeing the ImageDeleter attempt to list the snapshots on the local image (to delete), detect that it needs to delete a snapshot, attempt to acquire the exclusive lock, and fail opening the journal since the local image is not primary.  This causes the exclusive lock to be kept in a locked state but the image is closed.  When ImageDeleter retries, it sees that the lock is till held (by itself) and causes the blacklist.

I'll fix the issue where the lock is not properly released upon journal error.  I think we will also need to expand the ```exclusive_lock::Policy``` to prevent opening the journal when you attempt to unprotect or delete snapshots.  I can create a PR for that and you could just swap out ```ImageCtx::exclusive_lock_policy``` with a new one that disables opening the journal.

rjfd commented:
@dillaman That is a good idea! I'll add more unit test to cover that case.

dillaman commented:
@rjfd Actually -- in the short term, you should be able to directly unset the RBD_FEATURE_JOURNALING bit after opening the image but before proceeding with snapshot manipulation.  Looking at adjusting the policy to support this looks like it might be more effort than it really solves.

rjfd commented:
@dillaman pushed the changes and rebased with master
Added more unit tests to test the cases of images with snapshots and clones.

rjfd commented on src/tools/rbd_mirror/ImageReplayer.cc:
```diff
@@ -404,6 +404,9 @@ void ImageReplayer<I>::handle_bootstrap(int r) {
 
   {
     Mutex::Locker locker(m_lock);
+    m_name = m_local_ioctx.get_pool_name() + "/" + m_local_image_ctx->name;
```
TODO: remove this line

dillaman commented:
@rjfd Hopefully last thing for this PR: the 'qa/workunits/rbd/rbd_mirror.sh' test needs to be updated because it currently fails.  It would be good to also add a simple deletion propagation test to that file.

rjfd commented:
@dillaman added a test case for the image deletion propagation. Also fixed a problem with the disabling of mirroring for non-primary images in commit c429f61 .

rjfd commented:
@dillaman Just rebased on top of master to resolve the conflicts. Also added one more bugfix to 48f005f

dillaman commented:
@rjfd thx -- I am going to build a test branch this afternoon with all the 'wip-jason-testing' bugs.

dillaman commented:
http://qa-proxy.ceph.com/teuthology/jdillaman-2016-05-22_12:32:14-rbd-wip-jd-testing---basic-mira/207665/teuthology.log

Invalid memory access detected in TestImageDeleter.Delete_NonPrimary_Image test

dillaman commented:
Resolved invalid memory access and merged manually
